src/Language/PureScript/Bridge/Printer.hs

```diff
@@ -96,31 +97,33 @@ sumTypeToText st =
     sep = T.replicate 80 "-"
 
 sumTypeToTypeDecls :: SumType 'PureScript -> Text
-sumTypeToTypeDecls st@(SumType t cs) = T.unlines $
+sumTypeToTypeDecls st@(SumType t cs _) = T.unlines $
     dataOrNewtype <> " " <> typeInfoToText True t <> " ="
   : "    " <> T.intercalate "\n  | " (map (constructorToText 4) cs) <> "\n"
-  : "derive instance generic" <> _typeName t <> " :: " <> genericConstraints <> genericInstance t <> "\n"
-  : [ "derive instance newtype" <> _typeName t <> " :: " <> newtypeInstance t <> " _\n" | isNewtype cs]
+  : instances st
   where
-    genericInstance = ("Generic " <>) . typeInfoToText False
-    newtypeInstance = ("Newtype " <>) . typeInfoToText False
-    genericConstraints
-        | stpLength == 0 = mempty
-        | otherwise = (<> " => ") $
-            if stpLength == 1
-                then genericConstraintsInner
-                else bracketWrap genericConstraintsInner
-    genericConstraintsInner = T.intercalate ", " $ map genericInstance sumTypeParameters
-    stpLength = length sumTypeParameters
-    bracketWrap x = "(" <> x <> ")"
-    sumTypeParameters = filter isTypeParam . Set.toList $ getUsedTypes st
-    isTypeParam typ = _typeName typ `elem` map _typeName (_typeParameters t)
-    isNewtype [constr]
-      | either isSingletonList (const True) (_sigValues constr) = True
-    isNewtype _   = False
-    dataOrNewtype = if isNewtype cs then "newtype" else "data"
-    isSingletonList [_] = True
-    isSingletonList _   = False
+    dataOrNewtype = if isJust (nootype cs) then "newtype" else "data"
+
+-- | Given a Purescript type, generate `derive instance` lines for typeclass
+-- instances it claims to have.
+instances :: SumType 'PureScript -> [Text]
+instances st@(SumType t _ is) = map go is
+  where
+    go :: Instance -> Text
+    go i = "derive instance " <> T.toLower c <> _typeName t <> " :: " <> extras i <> c <> " " <> typeInfoToText False t <> postfix i
+      where c = T.pack $ show i
```

Comment:
Well ok - if we rename the Constructors, this trick is obviously no longer possible, well it is, just the prefix would need to be stripped away.